packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -104,34 +102,52 @@ contract MixinExchangeCore is
             expirationTimestampInSec: orderValues[4],
             orderHash: getOrderHash(orderAddresses, orderValues)
         });
-
-        require(order.taker == address(0) || order.taker == msg.sender);
-        require(order.makerTokenAmount > 0 && order.takerTokenAmount > 0 && fillTakerTokenAmount > 0);
+        
+        // Validate order and maker only if first time seen
+        if (filled[order.orderHash] == 0 && cancelled[order.orderHash] == 0) {
+            require(order.makerTokenAmount > 0);
+            require(order.takerTokenAmount > 0);
+            require(isValidSignature(
+                order.orderHash,
+                order.maker,
+                signature
+            ));
+        }
+        
+        // Validate taker
+        if (order.taker != address(0)) {
+            require(taker == order.taker);
+        } else {
+            order.taker = taker;
+        }
         require(isValidSignature(
-            order.maker,
-            order.orderHash,
-            v,
-            r,
-            s
+            order.orderHash ^ 0x1, // Domain separator maker/taker
```

Comment:
If we have a different message format or add a salt, we can probably remove the separator.